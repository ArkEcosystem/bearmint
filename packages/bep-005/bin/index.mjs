#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
