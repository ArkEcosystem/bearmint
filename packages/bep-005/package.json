{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
