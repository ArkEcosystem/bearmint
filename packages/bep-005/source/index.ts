export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
