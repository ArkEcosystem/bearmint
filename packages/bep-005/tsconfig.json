{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
