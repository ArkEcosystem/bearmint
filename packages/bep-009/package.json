{
	"name": "@bearmint/bep-009",
	"version": "0.0.0",
	"description": "This library is the official reference implementation for BEP-009 (https://bep.dev/009)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"files": [
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch",
		"docs": "typedoc source",
		"knip": "knip",
		"lint": "eslint 'source/**/*.ts' --format=codeframe --config=../../config/eslintrc.js --fix",
		"test": "vitest run",
		"test:coverage": "vitest run --coverage",
		"test:ui": "vitest --ui",
		"test:watch": "vitest"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@ethereumjs/rlp": "^4.0.0",
		"@noble/hashes": "1.1.3",
		"camelcase": "^7.0.0",
		"camelcase-keys": "^8.0.2",
		"dayjs": "1.11.6",
		"dot-prop": "^7.2.0",
		"env-paths": "^3.0.0",
		"matcher": "^5.0.0",
		"moize": "6.1.3",
		"semver": "7.3.8",
		"snake-case": "^3.0.4",
		"super-regex": "^0.2.0"
	},
	"devDependencies": {
		"@bearmint/bep-005": "workspace:^",
		"@types/node": "18.11.9",
		"@types/semver": "7.3.13"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "public"
	}
}
