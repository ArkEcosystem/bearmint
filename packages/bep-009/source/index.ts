export * from './assert.js'
export * from './base64.js'
export * from './big-number.js'
export * from './bytes.js'
export * from './camel-case.js'
export * from './directory.js'
export * from './dot-property.js'
export * from './environment-paths.js'
export * from './file.js'
export * from './glob.js'
export * from './hash.js'
export * from './hex.js'
export * from './is.js'
export * from './json.js'
export * from './non-nullable-map.js'
export * from './number.js'
export * from './numbers.js'
export * from './package.js'
export * as Paths from './paths.js'
export * from './regex.js'
export * as RLP from './rlp.js'
export * from './semver.js'
export * from './snake-case.js'
export * from './string.js'
export * from './uniq.js'
export * from './zeit.js'
