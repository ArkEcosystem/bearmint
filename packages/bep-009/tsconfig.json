{
	"compilerOptions": {
		"outDir": "distribution",
		"typeRoots": ["node_modules/@types"]
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
