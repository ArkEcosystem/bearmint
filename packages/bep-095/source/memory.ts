import { hexToBytes, zeit } from '@bearmint/bep-009'
import { PROTO_OPTS } from '@bearmint/bep-013'
import { SignedTx } from '@bearmint/bep-018'

import type { Signature, StoreTx } from './database.js'
import type { MuSig } from './types.js'

type TxIndex = Record<string, string[]>

interface SignatureRequest {
	hash: string
	muSig: MuSig
	signature: Signature
}

export function makeMemory({ pendingLimit }: { pendingLimit: number }) {
	const txs: Record<string, StoreTx> = {}
	const txIdsBySender: TxIndex = {}
	const txIdsByPublicKey: TxIndex = {}

	function restoreTx(tx: StoreTx): string {
		const { sender } = SignedTx.fromBinary(hexToBytes(tx.hash), PROTO_OPTS)

		const pendingCount: number = (txIdsBySender[sender] ?? []).length

		if (pendingCount >= pendingLimit) {
			throw new Error(
				`The public key [${sender}] has reached its max of ${pendingLimit} pending txs.`,
			)
		}

		txs[tx.hash] = tx

		if (txIdsBySender[sender] === undefined) {
			txIdsBySender[sender] = []
		}

		txIdsBySender[sender]?.push(tx.hash)

		for (const publicKey of tx.muSig.slaves) {
			txIdsByPublicKey[publicKey] = txIdsByPublicKey[publicKey] ?? []
			txIdsByPublicKey[publicKey]?.push(tx.hash)
		}

		return tx.hash
	}

	function getTxById(storeId: string) {
		return txs[storeId]!
	}

	function removeById(storeId: string): void {
		const { hash, muSig } = txs[storeId]!
		const { sender } = SignedTx.fromBinary(hexToBytes(hash), PROTO_OPTS)

		// removes indexes
		txIdsBySender[sender] = txIdsBySender[sender]?.filter((id) => id !== storeId) ?? []

		for (const publicKey of muSig.slaves) {
			txIdsByPublicKey[publicKey] =
				txIdsByPublicKey[publicKey]?.filter((id) => id !== storeId) ?? []
		}

		// remove actual tx
		delete txs[storeId]
	}

	return {
		deleteAllTxs() {
			for (const id of Object.keys(txs)) {
				removeById(id)
			}
		},
		getAllTxs() {
			return Object.values(txs)
		},
		getTxById,
		getTxStoreByPublicKey(publicKey: string) {
			const storeIdsBySender = txIdsBySender[publicKey] ?? []
			const storeIdsByPublicKey = txIdsByPublicKey[publicKey] ?? []

			const allById: Record<string, StoreTx> = {}
			for (const id of [...storeIdsBySender, ...storeIdsByPublicKey]) {
				allById[id] = getTxById(id)
			}

			return Object.values(allById)
		},
		loadTxs(storeTxs: StoreTx[]) {
			for (const tx of storeTxs) {
				restoreTx(tx)
			}
		},
		removeById,
		saveTx({ hash, muSig, signature }: SignatureRequest): string {
			return restoreTx({
				hash,
				muSig,
				signatures: [signature],
				timestamp: zeit.toUnix(),
			})
		},
		updateTx({ hash, signature }: SignatureRequest) {
			txs[hash]?.signatures.push(signature)
		},
	}
}

export type Memory = ReturnType<typeof makeMemory>
