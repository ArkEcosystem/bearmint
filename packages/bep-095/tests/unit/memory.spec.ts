import { expect, describe, it, beforeEach, afterEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import { bytesToHex } from '@bearmint/bep-009'
import { SignedTx } from '@bearmint/bep-018'
import MockDate from 'mockdate'

import type { Memory } from '../../source/memory.js'
import { makeMemory } from '../../source/memory.js'
import { accounts, data, muSig } from '../fixture.js'

describe<{ hash: string; subject: Memory }>('Memory', () => {
	beforeEach((context) => {
		context.hash = bytesToHex(new SignedTx(data).toBinary())
		context.subject = makeMemory({ pendingLimit: 3 })

		MockDate.set(Date.parse('2022-05-09T13:14:01+00:00'))
	})

	afterEach(() => {
		MockDate.reset()
	})

	it('should create a tx with a signature and then append further signature', async (context) => {
		expect(context.subject.getAllTxs()).toHaveLength(0)

		context.subject.saveTx({
			hash: context.hash,
			muSig,
			signature: accounts[0],
		})

		expect(context.subject.getAllTxs()).toMatchSnapshot()

		context.subject.updateTx({
			hash: context.hash,
			muSig,
			signature: accounts[1],
		})

		expect(context.subject.getAllTxs()).toMatchSnapshot()

		context.subject.updateTx({
			hash: context.hash,
			muSig,
			signature: accounts[2],
		})

		expect(context.subject.getAllTxs()).toMatchSnapshot()
	})

	it('should throw if the max tx limit has been reached', async (context) => {
		context.subject.saveTx({
			hash: 'a',
			muSig,
			signature: accounts[0],
		})

		context.subject.saveTx({
			hash: 'b',
			muSig,
			signature: accounts[0],
		})

		context.subject.saveTx({
			hash: 'c',
			muSig,
			signature: accounts[0],
		})

		expect(() =>
			context.subject.saveTx({
				hash: 'd',
				muSig,
				signature: accounts[0],
			}),
		).toThrow('has reached its max of 3 pending txs')
	})

	it('should remove all txs', async (context) => {
		expect(context.subject.getAllTxs()).toHaveLength(0)

		context.subject.saveTx({
			hash: context.hash,
			muSig,
			signature: accounts[0],
		})

		expect(context.subject.getAllTxs()).toMatchSnapshot()

		context.subject.deleteAllTxs()

		expect(context.subject.getAllTxs()).toHaveLength(0)
	})

	it('should restore all txs', async (context) => {
		context.subject.saveTx({
			hash: context.hash,
			muSig,
			signature: accounts[0],
		})

		expect(context.subject.getAllTxs()).toMatchSnapshot()

		const txs = context.subject.getAllTxs()

		context.subject.deleteAllTxs()

		expect(context.subject.getAllTxs()).toHaveLength(0)

		context.subject.loadTxs(txs)

		expect(context.subject.getAllTxs()).toMatchSnapshot()
	})

	it('should restore all txs', async (context) => {
		expect(context.subject.getTxStoreByPublicKey(muSig.master)).toHaveLength(0)

		context.subject.saveTx({
			hash: context.hash,
			muSig,
			signature: accounts[0],
		})

		expect(context.subject.getTxStoreByPublicKey(muSig.master)).toMatchSnapshot()
	})
})
