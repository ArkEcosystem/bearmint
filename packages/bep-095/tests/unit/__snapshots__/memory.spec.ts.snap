// Vitest Snapshot v1

exports[`Memory > should create a tx with a signature and then append further signature 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should create a tx with a signature and then append further signature 2`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
      {
        "id": "09bba3b4ea213f652d4718747622aeb2e17846c79ed8bc15a6ea18135f94e4e3",
        "privateKey": "2755b1a9723ab1be5ea841c6270054f5d3f48322f76c9d1c9225d18c0c824d94",
        "publicKey": "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "signature": "af0448ced582139acffa9ef65db6a8f4cbcd473dab7c818dd888fc336eca7544d7436012acd6863ae62dd86123044ca3080d6c6b1034f4de6bb0578be0b945335821f592fffb197fb8c38ff1e9ec4bd4fe04427cc58a511b6e581db310ee7fb6",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should create a tx with a signature and then append further signature 3`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
      {
        "id": "09bba3b4ea213f652d4718747622aeb2e17846c79ed8bc15a6ea18135f94e4e3",
        "privateKey": "2755b1a9723ab1be5ea841c6270054f5d3f48322f76c9d1c9225d18c0c824d94",
        "publicKey": "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "signature": "af0448ced582139acffa9ef65db6a8f4cbcd473dab7c818dd888fc336eca7544d7436012acd6863ae62dd86123044ca3080d6c6b1034f4de6bb0578be0b945335821f592fffb197fb8c38ff1e9ec4bd4fe04427cc58a511b6e581db310ee7fb6",
      },
      {
        "id": "267d857c3a01f739055a34028f1c2051769b0ead4ef33a0eeea06be81a8ba503",
        "privateKey": "36aa2cd61677af51081c07166eca89a2117b2579a4efb26ae8f18fa7d69b7c99",
        "publicKey": "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
        "signature": "b365f6cb603596a23520f3875659aa2d59a3af115f73c6479343a6b37b84bd92ebfe42db46e335dd058a809df18999c717bdc9dbba49aeb894e52e17517dee6ee9a268665323edaeec32bd5fde1f351ba99feacbd8e9444f43643f59e0a3d867",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should remove all txs 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should restore all txs 1`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should restore all txs 2`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;

exports[`Memory > should restore all txs 3`] = `
[
  {
    "hash": "0a05312e302e3012606161646437613638366137343034383737633338663064333433373136663864316337316165383531353930323438353333353232386232353936383966313663353036643663353731323238633234666365656234326362376632666337621a09726563697069656e7420012801321b0a0768616e646c6572120776657273696f6e1a076e6574776f726b",
    "muSig": {
      "m": 2,
      "master": "aadd7a686a7404877c38f0d343716f8d1c71ae8515902485335228b259689f16c506d6c571228c24fceeb42cb7f2fc7b",
      "n": 3,
      "slaves": [
        "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "93e48f0084f3d3ac1c2c58a0f0e029d697f183b1d53f22c9cb0689d1448f7f6a0fcd8c6f7703c5b378392338bdee6ee8",
        "a0b72896d656c9a73ba34c2d38df03bf53257249f9515f57fa4276b838221a1e3aebde3f31d172ecac245220eef6c8ec",
      ],
    },
    "signatures": [
      {
        "id": "4bf94d385ec00d23a652717e384e1856d267dc8469092cd0d0d7b992a936976a",
        "privateKey": "0af283be3ba696060636c0790d967380e8aff1e46d3d50c54fbd035b5ce3fa11",
        "publicKey": "acaddab32e508845e601f6d67fce8d525b208fbcb8c437019cbff4154002cd0fd428497db89e2fae419390038e7bb9b6",
        "signature": "928564eb2358b0b30836dd6ca6e60c4ead04e914933ace89196417d57c39e31ffcea381ad686f6bb52d8f9833d4da933084103e7d1717d9863122488f2b63ed106906250d08a00f95a4eeedf57ff86aa7388efa61e94c311ca7d1e70fbfe6375",
      },
    ],
    "timestamp": 1652102041,
  },
]
`;
